src/GitHub.InlineReviews/Glyph/GlyphMargin.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Windows;
+using System.Windows.Media;
+using System.Windows.Controls;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using GitHub.InlineReviews.Glyph.Implementation;
+
+namespace GitHub.InlineReviews.Glyph
+{
+    public sealed class GlyphMargin<TGlyphTag> : IWpfTextViewMargin, ITextViewMargin, IDisposable where TGlyphTag: ITag
+    {
+        bool handleZoom;
+        bool isDisposed;
+        Grid marginVisual;
+        double marginWidth;
+        bool refreshAllGlyphs;
+        ITagAggregator<TGlyphTag> tagAggregator;
+        IWpfTextView textView;
+        string marginName;
+        GlyphMarginVisualManager<TGlyphTag> visualManager;
+        Func<ITextBuffer, bool> isMarginVisible;
+
+        public GlyphMargin(
+            IWpfTextViewHost wpfTextViewHost,
+            IGlyphFactory<TGlyphTag> glyphFactory,
+            Func<Grid> gridFactory,
+            ITagAggregator<TGlyphTag> tagAggregator,
+            IEditorFormatMap editorFormatMap,
+            Func<ITextBuffer, bool> isMarginVisible,
+            string marginPropertiesName, string marginName, bool handleZoom = true, double marginWidth = 17.0)
+        {
+            textView = wpfTextViewHost.TextView;
+            this.tagAggregator = tagAggregator;
+            this.isMarginVisible = isMarginVisible;
+            this.marginName = marginName;
+            this.handleZoom = handleZoom;
+            this.marginWidth = marginWidth;
+
+            marginVisual = gridFactory();
+            marginVisual.Width = marginWidth;
+
+            visualManager = new GlyphMarginVisualManager<TGlyphTag>(textView, glyphFactory, marginVisual,
+                this, editorFormatMap, marginPropertiesName);
+
+            // Do on Loaded to give diff view a chance to initialize.
+            marginVisual.Loaded += OnLoaded;
+        }
+
+        public void Dispose()
+        {
+            if (!isDisposed)
+            {
+                tagAggregator.Dispose();
+                marginVisual = null;
+                isDisposed = true;
+            }
+        }
+
+        public ITextViewMargin GetTextViewMargin(string marginName)
+        {
+            if (marginName == this.marginName)
+            {
+                return this;
+            }
+
+            return null;
+        }
+
+        void OnLoaded(object sender, RoutedEventArgs e)
+        {
+            RefreshMarginVisibility();
+
+            tagAggregator.BatchedTagsChanged += OnBatchedTagsChanged;
+            textView.LayoutChanged += OnLayoutChanged;
+            if (handleZoom)
```

Comment:
Maybe if the margin was being reused as a map view (under the scroll bar)? It isn't being used at the moment. I'm in two minds about removing it.